File: Combination Sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()
        Solution.res = []
        self.curse(candidates, [], 0, target)
        return Solution.res
        
    def curse(self, candidates, rest, start_idx, target):
        if target == 0:
            Solution.res.append(rest)
            return
        for i in range(start_idx, len(candidates)):
            if i > start_idx and candidates[i-1] == candidates[i]: continue
            elif candidates[i] <= target:
                self.curse(candidates, rest + [candidates[i]], i, target-candidates[i])
        

File: Combination Sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<int> rest;
        curse(candidates, rest, 0, target);
        return res;
    }
private:
    vector<vector<int>> res;
    void curse(vector<int>& candidates, vector<int>& rest, unsigned start_idx, int target){
        if (target == 0){
            res.push_back(rest);
            return;
        }
        for (unsigned i = start_idx; i < candidates.size(); i++){
            if (i > start_idx && candidates[i] == candidates[i-1]) continue;
            if (candidates[i] <= target){
                rest.push_back(candidates[i]);
                curse(candidates, rest, i, target-candidates[i]);
                rest.pop_back();
            }
        }
        return;
    }
};